### PR TITLE
add local affinity support for group router

### PR DESCRIPTION
**Pull request by pxsdirac**

### CODE REVIEW

**Comment by lightbend-cla-validator** [comment]
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

**Review by raboof — COMMENTED** [review]
This is promising, thanks for looking into it!

I think as it is proposed now it is a bit to strong though: if there are no local listeners, but there are remote listeners, I think a remote listener should be picked.

If you strongly want to route only to local actors, I think it would be easier to just do local routing without any knowledge of actors on other nodes.

**Comment by johanandren** [comment]
Refs #28174

**Comment by pxsdirac** [comment]
> This is promising, thanks for looking into it!
> 
> I think as it is proposed now it is a bit to strong though: if there are no local listeners, but there are remote listeners, I think a remote listener should be picked.
> 
> If you strongly want to route only to local actors, I think it would be easier to just do local routing without any knowledge of actors on other nodes.

yes, I have the same concern.

File: akka-actor-typed/src/main/scala/akka/actor/typed/scaladsl/Routers.scala
Comment (johanandren) [inline akka-actor-typed/src/main/scala/akka/actor/typed/scaladsl/Routers.scala]: It shouldn't be an optional parameter here but instead default to false and use the `GroupRouter` API to allow for changing it. Default parameters must be avoided in public APIs as they are close to impossible to evolve in a binary compatible way.

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/routing/GroupRouterImpl.scala
Comment (johanandren) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/routing/GroupRouterImpl.scala]: How it should work is that having `localAffinity = true` makes the router prefer local routees, but use remote if there are no local.

Comment (pxsdirac) [reply akka-actor-typed/src/main/scala/akka/actor/typed/internal/routing/GroupRouterImpl.scala]: so as raboof said. maybe I can use remote actorref if there is no local actor ref, I will fix it.

Comment (johanandren) [reply akka-actor-typed/src/main/scala/akka/actor/typed/internal/routing/GroupRouterImpl.scala]: Yes, exactly 👍

Comment (johanandren) [reply akka-actor-typed/src/main/scala/akka/actor/typed/internal/routing/GroupRouterImpl.scala]: Also , note that it already does some come complicated logic here with preferring reachable over all registered actors for the service key, so it will need to be a three step logic with local, if no local reachable, if no reachable unreachable routees.

**Comment by lightbend-cla-validator** [comment]
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

**Comment by lightbend-cla-validator** [comment]
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/routing/GroupRouterImpl.scala
Comment (pxsdirac) [reply akka-actor-typed/src/main/scala/akka/actor/typed/internal/routing/GroupRouterImpl.scala]: maybe I should dive into the code and think more about it.

**Review by johanandren — COMMENTED** [review]
I think that's how the logic should be, please add some test coverage, it will need to live in the `akka-cluster-typed` tests to actually use cluster.

**Comment by johanandren** [comment]
Also, note that you will need to make sure you hav the same email in the commits as you have signed the Lightbend CLA, you can do that globally with `git config --global user.email "john@doe.org"` or just for your clone/fork of akka with `git config user.email "john@doe.org"`

File: akka-actor-typed/src/main/scala/akka/actor/typed/scaladsl/Routers.scala
Comment (patriknw) [inline akka-actor-typed/src/main/scala/akka/actor/typed/scaladsl/Routers.scala]: Maybe `localAffinity` parameter should be on the specific routing logic instead on all group routers level? For example I don't think it makes sense to allow localAffinity for consistent hashing.

The api could be in the `withRandomRouting` and `withRoundRobinRouting` methods instead.
```
Routers.group(key).withRoundRobinRouting(localAffinity = true)
```

**Review by patriknw — COMMENTED** [review]
looking good, but one thought about the api

File: akka-actor-typed/src/main/scala/akka/actor/typed/scaladsl/Routers.scala
Comment (johanandren) [reply akka-actor-typed/src/main/scala/akka/actor/typed/scaladsl/Routers.scala]: I wonder if it should be `preferLocalRoutees` as well, as that makes what it does more obvious and less academic-hard-wordy?

Comment (patriknw) [reply akka-actor-typed/src/main/scala/akka/actor/typed/scaladsl/Routers.scala]: sounds good to me

**Comment by pxsdirac** [comment]
what is done in the recent commit:
1. change the api,  now, the api is in `withRandomRouting` and `withRoundRobinRouting`. and rename the argument with `preferLocalRoutees`.
2. add unit tests for group router.
3. force push to clean the old commits so that CLA works.

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/routing/GroupRouterImpl.scala
Comment (patriknw) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/routing/GroupRouterImpl.scala]: isn't `preferLocalRoutees` already (always) `false` for consistent hashing and this pattern match isn't needed?

File: akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/Routers.scala
Comment (patriknw) [inline akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/Routers.scala]: would be nice to explain the `preferLocalRoutees` parameter

**Review by patriknw — COMMENTED** [review]
very good, just two small comments before my approval...

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/routing/GroupRouterImpl.scala
Comment (pxsdirac) [reply akka-actor-typed/src/main/scala/akka/actor/typed/internal/routing/GroupRouterImpl.scala]: you are right, but there is a place need to small fix:
in 
```
def withConsistentHashingRouting(virtualNodesFactor: Int, mapping: T => String): GroupRouterBuilder[T] = {
    import akka.actor.typed.scaladsl.adapter._
    copy(
      logicFactory = system =>
        new RoutingLogics.ConsistentHashingLogic[T](
          virtualNodesFactor,
          mapping,
          system.toClassic.asInstanceOf[ExtendedActorSystem].provider.getDefaultAddress))
  }
```
need to change to:
```
def withConsistentHashingRouting(virtualNodesFactor: Int, mapping: T => String): GroupRouterBuilder[T] = {
    import akka.actor.typed.scaladsl.adapter._
    copy(
      preferLocalRoutees = false, 
      logicFactory = system =>
        new RoutingLogics.ConsistentHashingLogic[T](
          virtualNodesFactor,
          mapping,
          system.toClassic.asInstanceOf[ExtendedActorSystem].provider.getDefaultAddress))
  }
```

Comment (pxsdirac) [reply akka-actor-typed/src/main/scala/akka/actor/typed/internal/routing/GroupRouterImpl.scala]: I will fix that.

**Comment by pxsdirac** [comment]
it seems binary compatibility checked failed. working on that.

**Comment by pxsdirac** [comment]
@patriknw I have no experience how to work with `backwards binary compatibility`. I add 
```
# Those are new methods required for prefer local routees.
ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.actor.typed.javadsl.GroupRouter.withRandomRouting")
ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.actor.typed.javadsl.GroupRouter.withRoundRobinRouting")
ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.actor.typed.scaladsl.GroupRouter.withRandomRouting")
ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.actor.typed.scaladsl.GroupRouter.withRoundRobinRouting")
ProblemFilters.exclude[DirectMissingMethodProblem]("akka.actor.typed.internal.routing.GroupRouterImpl.this")
ProblemFilters.exclude[DirectMissingMethodProblem]("akka.actor.typed.internal.routing.InitialGroupRouterImpl.this")
```
and it passes. but I don't know whether I should do it or not.

**Comment by johanandren** [comment]
`GroupRouter` is not for user extension, and the internals are safe to change bincomp for so please add a mima exclusions file for this!

**Comment by johanandren** [comment]
OK TO TEST

**Review by johanandren — APPROVED** [review]
LGTM, great work @pxsdirac !

**Comment by patriknw** [comment]
Code is looking good. Would be good to mention it in the documentation, in the section Routing Strategies in akka-docs/src/main/paradox/typed/routers.md

https://doc.akka.io/docs/akka/current/typed/routers.html#routing-strategies

After that we can merge

**Review by patriknw — APPROVED** [review]
LGTM

**Comment by patriknw** [comment]
Thanks for excellent contribution @pxsdirac 
It will be included in upcoming Akka 2.6.1.

**Pull request merged**
